doc: improve guidance for gitian lxc building

patricklodder opened the pull request:
This addition to the gitian doc explains some common troubleshooting when running gitian builder with LXC and how to adapt configuration to make the LXC container talk to the host.

chromatic commented:
Concept ACK. I tried these instructions and see:

```
--- Building for bionic amd64 ---
Stopping target if it is up
Making a new image copy
cp: cannot stat 'base-bionic-amd64': No such file or directory
Traceback (most recent call last):
	6: from ./bin/gbuild:348:in `<main>'
	5: from ./bin/gbuild:348:in `each'
	4: from ./bin/gbuild:350:in `block in <main>'
	3: from ./bin/gbuild:350:in `each'
	2: from ./bin/gbuild:355:in `block (2 levels) in <main>'
	1: from ./bin/gbuild:60:in `build_one_configuration'
./bin/gbuild:23:in `system!': failed to run make-clean-vm --suite bionic --arch amd64 (RuntimeError)
```

... though I think that's irrelevant to the new environment variables recommended here. If you can confirm my suspicions, I can approve this.

patricklodder commented:
I think it's because you need to run with `--lxc --setup` first? It should create `base-bionic-amd64` (an image file) in the working directory you're calling the script from.

chromatic commented:
Yes, that helped. I still have errors, but comparing with and without these environment variables, with is better.

patricklodder commented:
what kind of errors do you get? Would be good to cover them in the doc.

chromatic commented:
I set the LXC bridge and gateway environment variables appropriately.

```
$ ./gitian-build.sh --build --lxc -u https://github.com/patricklodder/dogecoin -c 1.14.4-libevent-2.1.12
From https://github.com/patricklodder/dogecoin
 * branch                1.14.4-libevent-2.1.12 -> FETCH_HEAD
--- Building for bionic amd64 ---
Stopping target if it is up
Making a new image copy
lxc-execute: gitian: conf.c: turn_into_dependent_mounts: 3915 No such file or directory - Failed to recursively turn old root mount tree into dependent mount. Continuing...
lxc-execute: gitian: sync.c: sync_wait: 34 An error occurred in another process (expected sequence number 4)
lxc-execute: gitian: start.c: __lxc_start: 2074 Failed to spawn container "gitian"
lxc-execute: gitian: tools/lxc_execute.c: main: 224 Failed run an application inside container
Traceback (most recent call last):
	6: from ./bin/gbuild:347:in `<main>'
	5: from ./bin/gbuild:347:in `each'
	4: from ./bin/gbuild:349:in `block in <main>'
	3: from ./bin/gbuild:349:in `each'
	2: from ./bin/gbuild:354:in `block (2 levels) in <main>'
	1: from ./bin/gbuild:60:in `build_one_configuration'
./bin/gbuild:23:in `system!': failed to run make-clean-vm --suite bionic --arch amd64 (RuntimeError)
```

xanimo reviewed (commented):
I've finally got it building past the previous errors I ran into today (with a simple [script](https://gist.githubusercontent.com/xanimo/4a95d483614cd4716386e636448f95f2/raw/d344ff66946c41e90c9be5a7825b1b8afe3eab90/lxc-gitian.sh)) but the key difference this time around is I'm running on Ubuntu 18.04. 

Will update progress later tonight if success, otherwise I'll keep triaging. Including a copy of my [experiments](https://gist.github.com/xanimo/6db99d8f801ff558d405016307833159) with focal and jammy for anticipated troubleshooting down the line.

Outside that, my only pertinent findings so far re document feedback are to add `debootstrap` to the list of essential packages for ubuntu.

patricklodder commented:
Thanks @xanimo. I also have spun up a Ubuntu Jammy box to be able to test there and I run into similar issues as you there, your script works for me on my Hirsute box too, but not on my Focal box after updating to the latest lxc (dumb move on my part)

| Ubuntu |  `lxc` | `debootstrap` | `apt-cacher-ng` | works? |
| :-: | :-: | :-: | :-: | :-: |
| 18.04 Bionic | `3.0.3-0ubuntu1` | `1.0.95ubuntu0.11` | `3.1-1build1` | ✅ |
| 20.04 Focal | `4.0.12-0ubuntu1` | `1.0.118ubuntu1.7` | `3.3.1-2build1` | ❌ |
| 21.04 Hirsute | `4.0.6-0ubuntu1` | `1.0.123ubuntu6.2` | `3.6.3-1` | ✅ |
| 21.10 Impish | `4.0.12-0ubuntu1` | `1.0.124ubuntu0.2` | `3.6.4-1` | ❌ |
| 22.04 Jammy | `5.0.0~git2209-g5a7b9ce67-0ubuntu1` | `1.0.126+nmu1ubuntu0.1` | `3.7.4-1build1` | ❌ |

Looking at this, there seems to be a regression between the `gitian-builder` script and lxc 4.0.6 and 4.0.12. @chromatic can you confirm that your lxc version is `> 4.0.6`? (`dpkg -s lxc | grep Version`)

I propose to add a note that LXC only works on a bionic host (as Hirsute is unmaintained) and then we look into this further next release cycle. Thoughts?

patricklodder commented:
I've added debootstrap, added the steps @xanimo and I took last night to get past a package install issue, and added the note that lxc only works on bionic.

chromatic commented:
Confirmed; I have `lxc` version 4.0.12 on the box where I see the errors.

chromatic commented on doc/gitian-building.md:
```diff
@@ -67,13 +67,75 @@ newgrp docker
 
 ### LXC
+
+**Please note that as of July 2022, the only host system that has been used to
+build successfully is Ubuntu Bionic (18.04). If your host is newer than that,
+it is easier to use Docker for building, see the section above.**
```
This note satisfies my concerns based on the results I've seen with `lxc`.

chromatic reviewed (dismissed):
With the current updates, ACK from me. Good to go.

xanimo commented:
Unfortunately I'm still having issues. Going to try again on a fresh VPS with 18.04. Currently getting:
```
--- Building for bionic amd64 ---
Stopping target if it is up
Making a new image copy
sudo: unable to resolve host gitian: Resource temporarily unavailable
Reading package lists... Done
Building dependency tree... Done
The following packages will be REMOVED:
  rsyslog*
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
After this operation, 1466 kB disk space will be freed.
(Reading database ... 18583 files and directories currently installed.)
Removing rsyslog (8.32.0-1ubuntu4) ...
invoke-rc.d: could not determine current runlevel
 * Stopping enhanced syslogd rsyslogd
   ...done.
(Reading database ... 18525 files and directories currently installed.)
Purging configuration files for rsyslog (8.32.0-1ubuntu4) ...
dpkg: warning: while removing rsyslog, directory '/etc/apparmor.d/local' not empty so not removed
Processing triggers for systemd (237-3ubuntu10) ...
Adding 'local diversion of /sbin/initctl to /sbin/initctl.distrib'
Adding 'local diversion of /usr/bin/ischroot to /usr/bin/ischroot.distrib'
Adding 'local diversion of /usr/sbin/policy-rc.d to /usr/sbin/policy-rc.d.distrib'
Starting target
Checking if target is up
Preparing build environment
Updating apt-get repository (log in var/install.log)
Installing additional packages (log in var/install.log)
Traceback (most recent call last):
	6: from ./bin/gbuild:347:in `<main>'
	5: from ./bin/gbuild:347:in `each'
	4: from ./bin/gbuild:349:in `block in <main>'
	3: from ./bin/gbuild:349:in `each'
	2: from ./bin/gbuild:354:in `block (2 levels) in <main>'
	1: from ./bin/gbuild:120:in `build_one_configuration'
./bin/gbuild:23:in `system!': failed to run on-target -u root -e DEBIAN_FRONTEND=noninteractive apt-get --no-install-recommends -y install curl g++-aarch64-linux-gnu g++-7-aarch64-linux-gnu gcc-7-aarch64-linux-gnu binutils-aarch64-linux-gnu g++-arm-linux-gnueabihf g++-7-arm-linux-gnueabihf gcc-7-arm-linux-gnueabihf binutils-arm-linux-gnueabihf g++-7-multilib gcc-7-multilib binutils-gold git-core pkg-config autoconf libtool automake faketime bison bsdmainutils ca-certificates python >> var/install.log 2>&1 (RuntimeError)
```

patricklodder commented:
Can you share your `var/install.log`?

xanimo commented:
> Can you share your `var/install.log`?

[var/install.log](https://gist.github.com/xanimo/88acfc32c07a3700e9985481c5c28360)
[var/build.log](https://gist.github.com/xanimo/ba8648259f8d77ceef8f35ecf30267c7)

xanimo commented:
> > Can you share your `var/install.log`?
> 
> [var/install.log](https://gist.github.com/xanimo/88acfc32c07a3700e9985481c5c28360) [var/build.log](https://gist.github.com/xanimo/ba8648259f8d77ceef8f35ecf30267c7)

I'm going to try building a commit from yesterday prior to merging #2409 because after more digging, it appears to be failing on libevent when building depends:
```
Fetching libevent-2.1.12-stable.tar.gz from https://github.com/libevent/libevent/releases/download/release-2.1.12-stable/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Transient problem: timeout Will retry in 1 seconds. 3 retries left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Transient problem: timeout Will retry in 2 seconds. 2 retries left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Transient problem: timeout Will retry in 4 seconds. 1 retries left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: github.com
Fetching libevent-2.1.12-stable.tar.gz from https://depends.dogecoincore.org
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Transient problem: timeout Will retry in 1 seconds. 3 retries left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Transient problem: timeout Will retry in 2 seconds. 2 retries left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Warning: Transient problem: timeout Will retry in 4 seconds. 1 retries left.
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: depends.dogecoincore.org
funcs.mk:244: recipe for target '/home/ubuntu/cache/common/download-stamps/.stamp_fetched-libevent-libevent-2.1.12-stable.tar.gz.hash' failed
make: *** [/home/ubuntu/cache/common/download-stamps/.stamp_fetched-libevent-libevent-2.1.12-stable.tar.gz.hash] Error 6
make: Leaving directory '/home/ubuntu/build/dogecoin/depends'

```
Will post an update shortly.

xanimo commented:
Am in the middle of my first actually running gitian build of dogecoin/1.14.6-dev with lxc. Attaching my notes for the evening: https://gist.github.com/xanimo/d8f3651ff583d75b4d4361b2169f3d39

@chromatic unaware if you were able to build using lxc but if not and you happen to have time I wonder if you or anyone that hasn't successfully built with lxc might be able to try building this commit before zlib was updated 3 months ago?

`./bin/gbuild --memory 14000 --commit dogecoin=c6a285828ea2c4d4ebcce5125c1077340be409cd ../dogecoin/contrib/gitian-descriptors/gitian-linux.yml`

I'm building manually from the gitian-builder directory but perhaps it'd work with the gitian-build.sh script. And sorry all for the essay.

patricklodder commented:
> I used gbuild manually.

What exact command line and env did you use?

patricklodder commented:
Update: I ran into the dns problem on the lxc guest machine too (on a completely fresh box+install with lxc 4.0.6) - so we'll have to put the lxc side of this back back to the drawing board.

chromatic commented:
@patricklodder You had an lxc build succeed in #3141 ; can we bring this PR in now?

patricklodder commented:
I had to fall back to using a bionic host because I still don't have a working environment on a recent Ubuntu.

patricklodder commented:
I'm not going to finish this, as gitian-builder is EOL. Will instead focus on guix for 1.15
